File: app/src/main/java/org/scp/gymlog/room/entities/BitEntity.java
package org.scp.gymlog.room.entities;

import static androidx.room.ForeignKey.CASCADE;
import static androidx.room.ForeignKey.SET_NULL;

import androidx.annotation.NonNull;
import androidx.room.Embedded;
import androidx.room.Entity;
import androidx.room.ForeignKey;
import androidx.room.Index;
import androidx.room.Junction;
import androidx.room.PrimaryKey;
import androidx.room.Relation;

import java.util.Date;
import java.util.List;

@Entity(
        tableName = "bit",
        foreignKeys = {
                @ForeignKey(
                        entity = ExerciseEntity.class,
                        parentColumns = "exerciseId",
                        childColumns = "exerciseId",
                        onDelete = CASCADE,
                        onUpdate = CASCADE),
                @ForeignKey(
                        entity = TrainingEntity.class,
                        parentColumns = "trainingId",
                        childColumns = "trainingId",
                        onDelete = CASCADE,
                        onUpdate = CASCADE),
        },
        indices = {
                @Index("exerciseId"),
                @Index("trainingId"),
                @Index({ "exerciseId", "timestamp" }),
                @Index({ "exerciseId", "trainingId", "timestamp" }),
        }
)
public class BitEntity {
    @PrimaryKey(autoGenerate = true)
    public int bitId;
    public int exerciseId;
    public int trainingId;
    public int reps;
    public int totalWeight;
    public boolean kilos;
    @NonNull
    public String note;
    @NonNull
    public Date timestamp;

    public static class WithBar {
        @Embedded
        public BitEntity bit;
        @Relation(
                parentColumn = "barId",
                entityColumn = "barId"
        )
        public BarEntity bar;
    }
}

File: app/src/main/java/org/scp/gymlog/room/entities/BitEntity.kt
package org.scp.gymlog.room.entities

import androidx.room.Embedded
import androidx.room.Entity
import androidx.room.ForeignKey
import androidx.room.Index
import androidx.room.Junction
import androidx.room.PrimaryKey
import androidx.room.Relation
import org.scp.gymlog.util.Constants
import org.scp.gymlog.util.JsonUtils.JsonFieldName
import org.scp.gymlog.util.JsonUtils.NoJsonify
import java.time.LocalDateTime

@Entity(
    tableName = "bit",
    foreignKeys = [
        ForeignKey(
            entity = VariationEntity::class,
            parentColumns = ["variationId"],
            childColumns = ["variationId"],
            onDelete = ForeignKey.CASCADE,
            onUpdate = ForeignKey.CASCADE),
        ForeignKey(
            entity = TrainingEntity::class,
            parentColumns = ["trainingId"],
            childColumns = ["trainingId"],
            onDelete = ForeignKey.CASCADE,
            onUpdate = ForeignKey.CASCADE),
    ],
    indices = [
        Index("trainingId"),
        Index("trainingId", "timestamp"),
        Index("variationId"),
        Index("variationId", "timestamp"),
        Index("variationId", "trainingId", "timestamp"),
    ]
)
class BitEntity {
    @NoJsonify
    @PrimaryKey(autoGenerate = true)
    var bitId = 0
    @JsonFieldName("v")
    var variationId = 0
    @JsonFieldName("t")
    var trainingId = 0
    var reps = 0
    @JsonFieldName("w")
    var totalWeight = 0
    var kilos = false
    var instant = false
    @JsonFieldName("s")
    var timestamp: LocalDateTime = Constants.DATE_ZERO
    var superSet = 0

    class BitEntityWithNotes {
        @Embedded
        lateinit var bit: BitEntity

        @NoJsonify
        @Relation(
            parentColumn = "bitId",
            entityColumn = "noteId",
            associateBy = Junction(BitNoteCrossRef::class)
        )
        var notes: List<NoteEntity> = emptyList()
    }
}